compiler/rustc_borrowck/src/diagnostics/conflict_errors.rs

```diff
@@ -1840,7 +1842,22 @@ impl<'cx, 'tcx> MirBorrowckCtxt<'cx, 'tcx> {
             let root_place =
                 self.prefixes(borrow.borrowed_place.as_ref(), PrefixSet::All).last().unwrap();
             let local = root_place.local;
-            match self.body.local_kind(local) {
+            let local_kind = self.body.local_kind(local);
+            if let LocalKind::Temp = local_kind {
+                // Mutable references to constants and functions will implicitly create
+                // temporaries, even though immutable references don't. We use these below to add a
+                // note about this.
+                match &self.body.local_decls[local].local_info {
+                    Some(box LocalInfo::FnDefRef) => {
+                        implicit_temporary = Some(("function", "function pointer"));
+                    }
+                    Some(box LocalInfo::ConstRef { .. }) => {
+                        implicit_temporary = Some(("constant", "value"));
+                    }
+                    _ => {}
+                }
+            }
+            match local_kind {
                 LocalKind::ReturnPointer | LocalKind::Temp => {
                     ("temporary value".to_string(), "temporary value created here".to_string())
```

Comment:
As you already mentioned, this message is the misleading part. Maybe pick a different message here for ConstRef and FnDefRef instead of adding a new note?